### DISCUSSION TITLE
Dismiss effect stops working depending on what I do in the parent reducer.

### DISCUSSION THREAD

**Discussion by SilverTab** [pointfreeco/swift-composable-architecture#2388] [Q&A] [answered]
I noticed a weird issue which I can't explain with the `dismiss` dependency. I have a simple project (see attached) with a parent reducer that shows a list, and when you select an item, it populates the optional "details" state, and presents the item's details view using drill-down navigation.

Whenever the details view is dismissed, I would like to perform an action on the parent reducer (in this case, re-fetch data from an API, and refresh the list of items). Everything works as expected when I dismiss the details view using the "Back" navigation button, but not when I try to use the `dismiss` dependency. When I do that, it seem like all the expected actions and state changes still happen, but the detail view remains presented and is never dismissed (even though its state is nil'ed out on the parent state).

It's much easier to demonstrate the issue with a sample project so I put together a quick example using the public Star Wars API that replicates the problem. Here's the relevant part of the parent reducer:

```
case .task:
    // This uses the shared "fetchData" function to refresh our data.
    return fetchData(state: &state)


// The details view is dismissed... refresh our data
case .planetDetails(.dismiss):
    // This works just fine
    return .run { send in
        await send(.task)
    }
    
    // this here, breaks the child's `dismiss` dependency
    // return fetchData(state: &state)

...

func fetchData(state: inout State) -> Effect<Action> {
    state.dataState = .loading
    return .run { send in
        await send(
            .planetListResponse(
                TaskResult {
                    try await starWarsClient.fetchPlanetList()
                }
            )
        )
    }
}
```

Sample project:
[StarWars.zip](https://github.com/pointfreeco/swift-composable-architecture/files/12396233/StarWars.zip)

**Comment by mbrandonw** [comment]
Hi @SilverTab, I looked into this briefly, and I'm not entirely sure it's a `\.dismiss` dependency problem. If you use `._printChanges()` on the root reducer then you will see that in the logs everything appears to be working correctly when the "Back" button is tapped. The `dismiss` action is received, the `dataState` goes to `.loading`, and `planetDetails` is `nil`'d out:

```diff
received action:
  PlanetListFeature.Action.planetDetails(
    .presented(.dismissTapped)
  )
  (No state changes)

received action:
  PlanetListFeature.Action.planetDetails(.dismiss)
  PlanetListFeature.State(
-   dataState: .success(
-     [
-       [0]: PlanetBrief(
-         name: "Tatooine",
-         url: "https://swapi.dev/api/planets/1/"
-       ),
-       [1]: PlanetBrief(
-         name: "Alderaan",
-         url: "https://swapi.dev/api/planets/2/"
-       ),
-       [2]: PlanetBrief(
-         name: "Yavin IV",
-         url: "https://swapi.dev/api/planets/3/"
-       ),
-       [3]: PlanetBrief(
-         name: "Hoth",
-         url: "https://swapi.dev/api/planets/4/"
-       ),
-       [4]: PlanetBrief(
-         name: "Dagobah",
-         url: "https://swapi.dev/api/planets/5/"
-       ),
-       [5]: PlanetBrief(
-         name: "Bespin",
-         url: "https://swapi.dev/api/planets/6/"
-       ),
-       [6]: PlanetBrief(
-         name: "Endor",
-         url: "https://swapi.dev/api/planets/7/"
-       ),
-       [7]: PlanetBrief(
-         name: "Naboo",
-         url: "https://swapi.dev/api/planets/8/"
-       ),
-       [8]: PlanetBrief(
-         name: "Coruscant",
-         url: "https://swapi.dev/api/planets/9/"
-       ),
-       [9]: PlanetBrief(
-         name: "Kamino",
-         url: "https://swapi.dev/api/planets/10/"
-       )
-     ]
-   ),
+   dataState: .loading,
-   _planetDetails: PlanetDetailsFeature.State(
-     planetURL: "https://swapi.dev/api/planets/1/",
-     dataState: .success(
-       PlanetInfo(
-         name: "Tatooine",
-         url: "https://swapi.dev/api/planets/1/",
-         climate: "arid",
-         population: "200000",
-         gravity: "1 standard",
-         terrain: "desert"
-       )
-     )
-   )
+   _planetDetails: nil
  )
```

If the screen isn't being popped off the screen then I am inclined to think that there is a SwiftUI bug lurking in the shadows here. After all, the state is definitely `nil`, and so now SwiftUI is in an inconsistent state. It's probably also possible to reproduce this in vanilla SwiftUI.

In general, `navigationDestination(isPresented:)` (which is what our `navigationDestination(store:)` uses under the hood) is quite buggy. I'm not sure there is much that can be done about it, other than using `NavigationStackStore` instead.

**Reply by SilverTab** [reply]
Thanks for looking into it! The weird thing is that, doing the exact same thing in a slightly different way (i.e.

```
return .run { send in
    await send(.task)
}
```
instead of just doing the shared work (i.e. `await return fetchData(state: &state)` ) seems to work just fine.  

> In general, navigationDestination(isPresented:) (which is what our navigationDestination(store:) uses under the hood) is quite buggy. I'm not sure there is much that can be done about it, other than using NavigationStackStore instead.

That's unfortunate! Does that imply adopting the [Stack-based](https://pointfreeco.github.io/swift-composable-architecture/1.0.0/documentation/composablearchitecture/stackbasednavigation) navigation style instead of the [Tree-based](https://pointfreeco.github.io/swift-composable-architecture/1.0.0/documentation/composablearchitecture/treebasednavigation) style?  I feel like the latter would suit the intended UI/UX much better, but if it's riddled with SwiftUI bug, I might have to re-consider 🤔

**Reply by mbrandonw** [reply]
> instead of just doing the shared work (i.e. `await return fetchData(state: &state)` ) seems to work just fine.
> 

Yeah, that is strange, but doing it that style does incur a thread hop, and so maybe that gives SwiftUI enough time to figure things out.

> That's unfortunate! Does that imply adopting the [Stack-based](https://pointfreeco.github.io/swift-composable-architecture/1.0.0/documentation/composablearchitecture/stackbasednavigation) navigation style instead of the [Tree-based](https://pointfreeco.github.io/swift-composable-architecture/1.0.0/documentation/composablearchitecture/treebasednavigation) style? I feel like the latter would suit the intended UI/UX much better, but if it's riddled with SwiftUI bug, I might have to re-consider 🤔

Yeah, exactly. Navigation stacks have been a lot less buggy than `navigationDestination`.

**Reply by SilverTab** [reply]
@mbrandonw huh, even weirder actually, just commenting the following line in the "shared computation" function fixes it, dismiss starts to work just fine. Just mentioning it in case you feel like it could move the blame from SwiftUI to TCA and could end up being useful (if not, ignore me 🙈)

```
func fetchData(state: inout State) -> Effect<Action> {
// commenting the following line fixes the dismiss issue
//    state.dataState = .loading
    return .run { send in
        await send(
            .planetListResponse(
                TaskResult {
                    try await starWarsClient.fetchPlanetList()
                }
            )
        )
    }
}
```

**Reply by mbrandonw** [reply]
Hi @SilverTab, here's another variation that somehow works with no changes to the reducer. Just replace your `PlanetListView` with this and continue using the `fetchData` method in the `.dismiss` action:

```swift
struct PlanetListView: View {
  let store: StoreOf<PlanetListFeature>
  
  var body: some View {
    NavigationStack {
      WithViewStore(self.store, observe: { $0 }) { (viewStore: ViewStoreOf<PlanetListFeature>) in
        Button {
          viewStore.send(.planetTapped("https://swapi.dev/api/planets/10/"))
        } label: {
          Text("Go")
        }
        .navigationDestination(
          store: self.store.scope(
            state: \.$planetDetails,
            action: { .planetDetails($0) }
          )
        ) { store in
          PlanetDetailsView(store: store)
        }
      }
    }
  }
}
```

Somehow that works fine. My only hunch is that somehow the `switch` at the root of the `NavigationStack` is tripping up SwiftUI somehow. This does make me think it is a SwiftUI bug and not a TCA bug.

**Reply by mbrandonw** [reply, marked as answer]
Hey @SilverTab, actually I just figured it out. It's this right here:

```swift
case let .success(planets):
  VStack(alignment: .leading) {
    …
  }
    .navigationDestination(
      store: self.store.scope(
        state: \.$planetDetails,
        action: { .planetDetails($0) }
      )
    ) { store in
      PlanetDetailsView(store: store)
    }
```

This applies the `navigationDestination` only to the `success` case, and so when you go back to `.loading` SwiftUI stops listening for that state change. You need to move that view modifier to a place where it is present in all cases of your enum. Either by moving the root view to a dedicated `RootView` kind of thing, or wrapping the whole thing in a `Group`.

So this is actually neither a bug in SwiftUI or TCA. 🙂

**Reply by SilverTab** [reply]
@mbrandonw Oh wow! Subtle! Happy to hear it's not another pesky SwiftUI navigation bug. Really appreciate you taking the time to look into this! 🙏

**Reply by jmprice18** [reply]
This has happened to me a few times and I did find navigationDestination documentation that supports this behavior. 

https://developer.apple.com/documentation/swiftui/view/navigationdestination(for:destination:)


> Do not put a navigation destination modifier inside a “lazy” container, like List or LazyVStack. These containers create child views only when needed to render on screen. Add the navigation destination modifier outside these containers so that the navigation stack can always see the destination.
